### ISSUE TITLE
i am warning you peepz

### ISSUE THREAD

**Issue by Andrews54757** [AJS-development/Project-Nehemiah#1]
I will be much much stricter here. one screw up=ban

**Comment by Andrews54757** [comment]
only issues allowed

**Comment by AlexandriteErikParacelsus** [comment]
K sir I will follow im skurred of u nao

**Comment by Andrews54757** [comment]
Do not comment anybody else

**Comment by LegitSoulja** [comment]
@Andrews54757 take it to the forums. I have resources aswell, where only trusted members are allowed to download this project. I thought this would be strict, so thats why I did that.

http://forum.ogarul.io < Login with your old account. Keep this repo hidden. It would be best to have in another location, unknown out of order name just for the download of this repo. Just keep this for updating, commits, etc. Where, from the forums you will download.

Trusted members can download by 2 ways.

* I thought about having them buy the rank, maybe not.
* Have them complete tasks such as (make a certain amount of posts, gain certain amounts of likes, or get all trophies).
* Have them be registered on the forums for a month before receiving? idk.

Saay something

**Comment by Andrews54757** [comment]
@LegitSoulja I already go the system down. (the key system) on my local computer. Trusted devs will have the keygen to create keys for others. Then, each of us can distribute it however we like. But I intend to keep it public on github so we can collaborate. So far, with this, I have created the basic mechanisms (food,server management,sockets,quadtree, and others) as well as a developing   console-only client. Currently, I need to create:


* a graphical client
* Moving nodes
* collisions
* viruses
* ejectedMass
* commands
* gamemodes
* plugins
* configs
* and more little tweaks

**Comment by LegitSoulja** [comment]
@Andrews54757 don't forget we have mysql databases if you want to experiment, and also sites if you need to use php system, with mysql. 

live.ogarul.io will be the domain/space for that content. I set up a POST system that receives the messages, and include whatever class. 

In any of those areas, if you need help let me know. 

btw, I did a vote for 2D or 3D, and 3D(4k) graphical settings took advantage. Should the client be 3D instead of 2D? I think with unity we can websockets (I think), but I know for sure with socket.io

How will the packets be handled? The same way? and node?

**Comment by Andrews54757** [comment]
The packets will be handled with events. Current packet events:


Client to server
* cha: chat packet. Uses RSON to stringify
* mouse: Mouse packet. Uses custom stringifying method
* mes: main message packet (play button, w, space, e,r,spectate,q)


Server to client:
* hello: A hello message. Happens when you join
* nodes: Nodes to be added. Uses RSON.
* delnodes: Nodes to be deleted. Uses RSON. (I need to implement this)
* rip: death of player (need to be implemented)
* reset: resets map (need to be implemented)



BTW, RSON is a object notation I built specifically for compact communication between string-only pipes that is able to handle almost anything (Includes strings, numbers,objects,arrays,functions, and circular data structures).

**Comment by Andrews54757** [comment]
Im thinking though, I might want someone to clean up my code for RSON and make it super efficient. Basically my goal for RSON was to solve the "JSON can't do circular references" error and other things. as well as make it considerably compact

**Comment by Andrews54757** [comment]
Sytax docs: https://github.com/AJS-development/RSON/wiki/Syntax

**Comment by LegitSoulja** [comment]
Clean it? Like make sure its right? work? or clean the code, smaller?

**Comment by Andrews54757** [comment]
Make it super efficient. That is my goal. I try to make it as efficient as possible, but I am bad at it

**Comment by LegitSoulja** [comment]
ah, i'll see. First I need to see how this works more specifically, clarity.

**Comment by Andrews54757** [comment]
Im putting some documentation comments for you

**Comment by LegitSoulja** [comment]
@Andrews54757 ok, I'll check it out

**Comment by LegitSoulja** [comment]
@Andrews54757 second thought, keep records of those people who did donate to charity water. They will have access to the downloads, without needing to obtain that membership on the forums.

All we need if you can get is there email, or some name identity, which you can search for on the forums(Admin panel). Make sure those users are added to the Trusted Members group.

And also, this RSON, I don't get it. Not the code, but the way it works. You say rson can store functions, but by looking at the output for recursion is what I don't understand. 

Basically, is there a more in dept description about how all of this works? Sorry, but it just gets a little frustrating with new terminology for a language, but you're most use to another.

**Comment by Andrews54757** [comment]
Sorry about that, like in the wiki "it is mostly only machine readable". But I will explain a bit later

**Comment by LegitSoulja** [comment]
@Andrews54757 Say for example, I created a branch, and you updated the main branch, whist I've made changes aswell to my branch. Is there a way to merge the main branch with mines? Just because, im not 100% sure how github works even if I did create a PR. I don't want to interrupt your current state in the project, but keep updated. You get what im saying?

**Comment by Andrews54757** [comment]
Yes, you use CLI. I can do that. But there are sometimes conflicts that are just really bad

**Comment by Andrews54757** [comment]
BTW, plugins will work off FlexiPlugins. Edit that : https://github.com/AJS-development/FlexiPlugins

**Comment by LegitSoulja** [comment]
@Andrews54757 oh, ok.

**Comment by Andrews54757** [comment]
@LegitSoulja . I have a question. In the new version, there is going to be a pointer pointing to the person who killed you. So that it makes the game more interesting. I got all the mechanisms working but I dont understand how to draw a "revenge icon" on the edge of the screen pointing to the person who killed you. Here is some code:

```
if (0 < playerCells.length) {
            calcViewZoom();
            var c = a = 0;
            for (var d = 0; d < playerCells.length; d++) {
                playerCells[d].updatePos();
                a += playerCells[d].x / playerCells.length;
                c += playerCells[d].y / playerCells.length;
            }
            posX = a;
            posY = c;
            posSize = viewZoom;
            nodeX = (nodeX + a) / 2;
            nodeY = (nodeY + c) / 2
        } else {
            nodeX = (29 * nodeX + posX) / 30;
            nodeY = (29 * nodeY + posY) / 30;
            viewZoom = (9 * viewZoom + posSize * viewRange()) / 10;
        }
 ctx.translate(canvasWidth / 2, canvasHeight / 2);
        ctx.scale(viewZoom, viewZoom);
        ctx.translate(-nodeX, -nodeY);

...
some more code (uneccesary)
...

if (killerId != -1 && players[killerId]) {
            var player = players[killerId];
           var center = player.calcCenter() // center of person who killed you
            if (center) {
             // what to do?
           
           
         
                
                
            } 
            
        } 
```

**Comment by Andrews54757** [comment]
You can get an idea of what i am trying to get by going over to http://wings.io

**Comment by LegitSoulja** [comment]
@Andrews54757 I don't see the idea over at wings.io, but I think I know what you mean

What you will need to do, is have a function which will update for however long until the dialog shows to join again.

You will have to draw a picture i guess, towards the players position, with a radius(How far from the player it should appear). Note the interval should be how many times per frame it should run, equivalent to the player speed.

Do you like need an example? or just a thought of doings?

**Comment by Andrews54757** [comment]
All I achieved was drawing a line to the killer : 
![screen shot 2016-10-10 at 8 02 18 pm](https://cloud.githubusercontent.com/assets/13282284/19254700/7e1ab64c-8f24-11e6-9851-48798796faa8.png)

**Comment by LegitSoulja** [comment]
@Andrews54757 
Ok thats good. 

Now, you need to translate, rotate, draw rectangle(transparent). Apply image, then set transform(Which will loop as a function)

**Comment by Andrews54757** [comment]
@LegitSoulja Why not? The new version can be unique. I was also thinking of special viruses that make you eject mass in all directions. like mitosis

**Comment by Andrews54757** [comment]
I will give you my client files through gitter

**Comment by LegitSoulja** [comment]
@Andrews54757 YES. Mitosis is the one that has all the crazy mass ejections with random patterns and shapes. I really like it. Beyond my abilities to create such a thing.

**Comment by Andrews54757** [comment]
No, thats petridish i think

**Comment by Andrews54757** [comment]
but I could also do that

**Comment by Andrews54757** [comment]
Because I am making a new move engine api which allows for objects to move to a certain location

**Comment by LegitSoulja** [comment]
oh, well if you can that would be amazing of you. That right there is what I call beauty.

**Comment by Andrews54757** [comment]
I think you should clean up the client a bit. Currently, the client lags, not the server

**Comment by LegitSoulja** [comment]
@Andrews54757 When i started to re-make the client, I rewrote it using es6. That's why I was kindof shock to see you working on it, but it still looks the same so I'll just apply the changes you made to the one I was working on. 

It's more flexible, and load time is faster.

**Comment by Andrews54757** [comment]
okay

**Comment by Andrews54757** [comment]
BTW. I still don't understand why you said that quad trees are better.

**Comment by LegitSoulja** [comment]
Because, i thought it would be more efficient. But I really don't know.

If you say you accomplished 500+ bots no lag then, i guess no worry. We're golden.

**Comment by Andrews54757** [comment]
To me these are the stats

| Action | Quadtree | Hashbounds |
|-------|-----------|--------------|
|Insert  | O(n)          | O(1), O(n) worst case|
|Delete | O(n)          | O(1), O(n) worst case|
|Get      | O(n)          | O(n), O(1) best case  |
| Space| O(log(n))  | O(n)  |

**Comment by Andrews54757** [comment]
I am also going to make every server run in its own child. But thats for later

**Comment by Andrews54757** [comment]
The tradeoff is that hashbounds is more efficient, yet consumes more memory

**Comment by Andrews54757** [comment]
@LegitSoulja I am also implementing a socket protection system. So this is how it works


1. Client connects
2. Server send its uid
3. Client uses an algorithm to create the key
4. The server checks if it is right
5. Otherwise the client is disconnected


But I am wondering what algorithm to use. I am thinking of using the crypto module

**Comment by LegitSoulja** [comment]
What's the point of doing this exactly?

Bot protection right? Only humans can access the client.

**Comment by Andrews54757** [comment]
@LegitSoulja I finished it. I will give you the updated client in gitter. You can test it yourself

**Comment by Andrews54757** [comment]
And you are right. It is for bot protection as well as very thin DDos protection. no one will notice it, as the system is very fast, and it can detect if it is fake in about 700 miliseconds

**Comment by LegitSoulja** [comment]
@Andrews54757 wonderful!. 

There's also solutions to ddos, like listening for multiple connections within a giving amount of seconds. Drop the connection if exceeded join time. This is what I was planning on implementing into ogarul, but maybe this is just better.

**Comment by Andrews54757** [comment]
I was also planning on adding kindof what you said. Dont accept connections if the server is busy.

**Comment by Andrews54757** [comment]
There is no real way to prevent DDos though. One way though is what you did. Dont tell people your ip address

**Comment by Andrews54757** [comment]
Mabye in the new version we could do that. Have a service that converts ips into our own "Domain names" where people can choose the name instead of their uid. And the only way to connect would then be via real clients

**Comment by LegitSoulja** [comment]
@Andrews54757 As long i've been a web developer, I've never really looked at, nor figured out how to allow people to create there own domains. 

For example, ogarul.io.. A server name is elite,
they will be able to setup **elite.ogarul.io** routing to there ip address.

I think since im using cloudflare, there's an api requests to do that. I'll look into it, that's if that what you mean by "convert ips to domain names"

**Comment by Andrews54757** [comment]
I mean connect it like so `play.ogarul.io/?ip=something.com`. We run the name service. And it is the same thing you did with statistics, except it is custom uids.

**Comment by LegitSoulja** [comment]
Ohhhh. Doh.

I getcha now.

**Comment by MasonBurdette** [comment]
@Andrews54757 will the Plugin API be opened for developers that would make plugins?  Also I'm wondering if there will be an API (HTTP Get) for the live server tracker.  Will unofficial clients be supported by the servers?

Anyways, the progress of this project looks great!  Btw I never got to help you guys contribute to the clean water charity, if you could send me the link, I'd love to help!

**Comment by Andrews54757** [comment]
Plugin api will be open, server api won't. And unofficial clients won't work because of the anti-bot/ddos protection system. It would be nice though, as the socket api for the client will be open. I will probably create a config to disable this protection, but note, without it, your server would be very vulnerable.

**Comment by LegitSoulja** [comment]
@MasonBurdette Due to the fact, the first release of the server tracker (Statistics), people was able to get json data from it, as to being able to get the list of servers.

Now, due to security, and privacy concerns it was removed, and won't be allowed anymore. Since the Live system(api) is secure, only to the network of ogarul.io, there's no way outside of the domain to get requests from our services.

Now, second note. You can make plugins, and also be able to use data for example as to mysql. With statistics, this same system is being used, but with extra security.

http://forum.ogarul.io/threads/live-api-server-side.47/

If you do, say for example want to use this you can, but if you do it will have to be moderated by me before uploading, due to the fact its PHP, which could potentially harm other things if it's not moderated. Or, host the php server side scripts yourself. But there will be less security.

Im pretty much providing a free databases for people who needs it.

**Comment by LegitSoulja** [comment]
@Andrews54757 What is v8-profiler? Module not found for me, while trying to run

**Comment by Andrews54757** [comment]
It allows for cpu sampling. Which I use to find out its efficiency and find performance leaks. to solve just do 

> npm install v8-profiler

**Comment by Andrews54757** [comment]
The data is recorded at ./data

**Comment by LegitSoulja** [comment]
@Andrews54757 After seeing my cpu go up and down rapidly, nothing really happened. I didn't get profile data. Files are empty, maybe not for windows?

And, is this project runnable yet?

**Comment by LegitSoulja** [comment]
@Andrews54757 I have no clue what's wrong with the sites.

I pretty much disabled all sites, but still seems as if somewhere there's a leak. Forums? Idk

**Comment by Andrews54757** [comment]
Take a cpu profile of your server and you will find out

**Comment by LegitSoulja** [comment]
@Andrews54757 

I have, but it's not showing specific information. I can't even figure out where the problems coming from. I disabled all domains, then everything calms down..

But right when xenforo gets added, it starts. What im thinking it's the robots, that's visiting the page. 

I also started adding limiters for the site. I'm trying everything I can, but right when I think its good, its not.

**Comment by LegitSoulja** [comment]
Going to try and remove the site from the search engine because I think it's a potential threat.

**Comment by LegitSoulja** [comment]
@Andrews54757 ok, everything seems to be good now. I've made a tester with nodejs to check the site every minute, and for the past hour it has been good.

So, therefor, the forums is now back up.
